Frameworks/Foundation/NSOperation.mm

```diff
@@ -131,139 +162,122 @@ - (void)setThreadPriority:(double)priority {
 */
-- (BOOL)isFinished {
-    return priv->finished != 0;
+- (BOOL)isExecuting {
+    return _executing;
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)isExecuting {
-    return priv->executing != 0;
+- (BOOL)isConcurrent {
+    return NO;
 }
 
 /**
  @Status Interoperable
 */
-- (void)start {
-    if (!priv->executing && !priv->finished) {
-        bool execute = false;
-
-        pthread_mutex_lock(&priv->finishLock);
-        if (!priv->cancelled) {
-            [self willChangeValueForKey:@"isExecuting"];
-            priv->executing = 1;
-            [self didChangeValueForKey:@"isExecuting"];
-            execute = true;
-        }
-        pthread_mutex_unlock(&priv->finishLock);
-        if (execute) {
-            [self main];
-        }
-
-        if (execute) {
-            [self willChangeValueForKey:@"isExecuting"];
-        }
-
-        [self _setFinished:YES
-            andPerformUnderLock:^{
-                if (execute) {
-                    priv->executing = 0;
-                    [self didChangeValueForKey:@"isExecuting"];
-                }
-            }];
-    }
+- (BOOL)isAsynchronous {
+    return NO;
 }
 
 /**
  @Status Interoperable
 */
-- (void)setFinished:(BOOL)finished {
-    [self _setFinished:finished andPerformUnderLock:nil];
-}
+- (void)start {
+    if (_finished) {
+        return;
+    }
 
-- (void)_setFinished:(BOOL)finished andPerformUnderLock:(void (^)())block {
-    // yes, this is ugly: priv->finished is an int though.
-    int newValue = finished ? 1 : 0;
-    pthread_mutex_lock(&priv->finishLock);
+    THROW_NS_IF(E_INVALIDARG, (_executing || ![self isReady]));
+    
+    _finishLock.lock();
+    BOOL shouldExecute = !_cancelled; // Note: in the cancelled case, [self main] is not called but isFinished will still be observable
+    if (shouldExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
+        _executing = YES;
+        [self didChangeValueForKey:@"isExecuting"];
+    }
+    _finishLock.unlock();
+    
+    if (shouldExecute) {
+        NSAutoreleasePool* pool = [NSAutoreleasePool new];
+        [self main];
+        [pool release];
+    }
 
-    [self willChangeValueForKey:@"isFinished"];
+    [self _finish:shouldExecute];
+}
 
-    if (priv->finished != newValue) {
-        priv->finished = newValue;
+- (void)_finish:(BOOL)didExecute {
+    _finishLock.lock();
+    if (didExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
     }
 
+    [self willChangeValueForKey:@"isFinished"];
+    _finished = YES;
     [self didChangeValueForKey:@"isFinished"];
 
-    if (block) {
-        block();
-    }
-
-    if (newValue == 1) {
-        pthread_cond_broadcast(&priv->finishCondition);
+    if (didExecute) {
+        _executing = NO;
+        [self didChangeValueForKey:@"isExecuting"];
     }
 
-    pthread_mutex_unlock(&priv->finishLock);
+    _completionBlockLock.lock();
+    void (^completion)(void) = [_completionBlock retain];
+    _completionBlock = nil;
```

Comment:
will be lot simpler to use completion = self.completionBlock
self.completionBlock = nil;
completion();


---
In reply to: [69981681](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69981681) [](ancestors = 69981681)